arangod/Aql/MaterializeExecutor.h

```diff
@@ -47,11 +48,36 @@ class RegisterInfos;
 template<BlockPassthrough>
 class SingleRowFetcher;
 
+// two storage varians as we need
+// collection name to be sotred only
+// when needed.
+class NoCollectionNameHolder {};
+class StringCollectionNameHolder {
+ public:
+  StringCollectionNameHolder(std::string const& name)
+      : _collectionSource(name) {}
+
+ protected:
+  std::string const& _collectionSource;
+};
+
 template<typename T>
-class MaterializerExecutorInfos {
+class MaterializerExecutorInfos
+    : private std::conditional_t<std::is_same_v<T, std::string const&>,
+                                 StringCollectionNameHolder,
+                                 NoCollectionNameHolder> {
  public:
-  MaterializerExecutorInfos(T collectionSource, RegisterId inNmDocId,
-                            RegisterId outDocRegId, aql::QueryContext& query);
+  using Base =
```

Comment:
If you write this using out of class you can use it twice instead of duplicate conditional_t